File: src/rules/seededRandom.js
// A seeded random number generator (SplitMix64) that the game rules can take in
// place of Math.random, so a rule run from a fixed seed is repeatable.
//
// Why this exists: the iOS app decides battle outcomes and prize draws on the
// device (ADR 0004), so the same rules live in JavaScript and in Swift. They are
// kept identical by golden files (ADR 0005): scripts/generate-golden.mjs runs the
// JavaScript rules from fixed seeds, and the Swift GameRules tests must reproduce
// the output exactly. That only works if both languages draw the same numbers,
// so the algorithm here is the reference — the Swift port must match it bit for
// bit, and golden/prng.json is the check that it does.
//
// SplitMix64 was picked for being tiny and easy to port exactly: one 64-bit
// state word, an add and three xor-shift-multiply steps, no tables. The 64-bit
// arithmetic uses BigInt because a JS number cannot hold it; BigInt.asUintN(64)
// is the wrapping `&+` / `&*` that Swift's UInt64 does natively.
//
// Plain ESM with no browser or Node APIs, so it bundles into the web app and
// can also be `require()`d from the CommonJS server on Node 22.12+.

const MASK = 64;
const GOLDEN_GAMMA = 0x9e3779b97f4a7c15n;
const MIX_1 = 0xbf58476d1ce4e5b9n;
const MIX_2 = 0x94d049bb133111ebn;
// 2^-53: the top 53 bits of an output, scaled, give every double in [0, 1)
// on an even 2^-53 grid — the same construction Swift uses for Double.random.
const FLOAT_SCALE = 2 ** -53;

// Accepts a non-negative safe integer, a BigInt, or a decimal/0x string, and
// wraps it to 64 bits, so a seed stored as a string in JSON round-trips.
function toSeed(seed) {
  if (typeof seed === 'number' && !Number.isSafeInteger(seed)) {
    throw new RangeError(`seed must be a safe integer, got ${seed}`);
  }
  return BigInt.asUintN(MASK, BigInt(seed));
}

// createSeededRandom(seed) → { next, nextUint64 }
//   next()        float in [0, 1), a drop-in for Math.random
//   nextUint64()  the raw 64-bit output as a BigInt, for exact comparison
// Both advance the same state, so interleaving them is still deterministic.
// The methods are closures, so `const rand = rng.next` can be passed around
// unbound wherever a rule expects a `() => number`.
export function createSeededRandom(seed) {
  let state = toSeed(seed);

  function nextUint64() {
    state = BigInt.asUintN(MASK, state + GOLDEN_GAMMA);
    let z = state;
    z = BigInt.asUintN(MASK, (z ^ (z >> 30n)) * MIX_1);
    z = BigInt.asUintN(MASK, (z ^ (z >> 27n)) * MIX_2);
    return z ^ (z >> 31n);
  }

  function next() {
    return Number(nextUint64() >> 11n) * FLOAT_SCALE;
  }

  return { next, nextUint64 };
}

File: ios/Packages/GameRules/Sources/GameRules/SeededRandom.swift
// The seeded random number generator (SplitMix64) the game rules draw from, so
// a rule run from a fixed seed is repeatable — the Swift port of
// src/rules/seededRandom.js.
//
// Draw-compatibility contract: the JavaScript generator is the reference, and
// this one must match it bit for bit (ADR 0005). Seeded with the same 64-bit
// value, the nth `nextUInt64()` here equals the nth `nextUint64()` there, and
// the nth `next()` equals the nth JS `next()` exactly — same Double, not merely
// close. `next()` is the top 53 bits of one 64-bit output times 2^-53, and each
// call to either method consumes exactly one output, so a rule that draws the
// same way in both languages sees the same numbers. golden/prng.json (written
// by `npm run golden:generate`) is the check; SeededRandomTests reads it.
//
// Rules therefore draw only through `RandomSource.next()`, in the order their
// JS twin documents, and never through the standard library's random APIs
// (`Int.random(in:using:)`, `shuffled(using:)`, …): those turn raw output into
// values with their own algorithms, which JavaScript doesn't reproduce. That's
// also why SeededRandom doesn't conform to RandomNumberGenerator.
//
// Only 64-bit wrapping arithmetic (`&+`, `&*`), which is what the JS side's
// BigInt.asUintN(64) emulates — no Foundation, like the rest of GameRules.

/// A source of floats in [0, 1) that rules take instead of reaching for a
/// global generator — the Swift form of the JavaScript rules' injected
/// `rng: () => number`. Pass a `SeededRandom` for repeatable results (tests,
/// golden parity) or a `SystemRandomSource` for live play, the way the web
/// falls back to `Math.random`.
public protocol RandomSource {
    /// The next float in [0, 1). Advances the source.
    mutating func next() -> Double
}

/// The top 53 bits of a 64-bit output times 2^-53: every value on an even
/// 2^-53 grid in [0, 1), exactly as JS computes `Number(x >> 11n) * 2 ** -53`.
/// Both steps are exact in Double, so there is no rounding to disagree on.
private func unitFloat(_ bits: UInt64) -> Double {
    Double(bits >> 11) * 0x1p-53
}

/// SplitMix64: one 64-bit state word, an add and three xor-shift-multiply
/// steps. A value type — copying it forks the sequence, so both copies draw the
/// same numbers from that point on.
public struct SeededRandom: RandomSource, Sendable, Equatable {
    private static let goldenGamma: UInt64 = 0x9e37_79b9_7f4a_7c15
    private static let mix1: UInt64 = 0xbf58_476d_1ce4_e5b9
    private static let mix2: UInt64 = 0x94d0_49bb_1331_11eb

    private var state: UInt64

    /// A fresh generator. Same seed as `createSeededRandom(seed)` in JS, same
    /// sequence.
    public init(seed: UInt64) {
        state = seed
    }

    /// The raw 64-bit output — for exact comparison against golden data.
    public mutating func nextUInt64() -> UInt64 {
        state = state &+ Self.goldenGamma
        var z = state
        z = (z ^ (z >> 30)) &* Self.mix1
        z = (z ^ (z >> 27)) &* Self.mix2
        return z ^ (z >> 31)
    }

    /// A float in [0, 1), a drop-in for `Math.random`; one 64-bit draw.
    public mutating func next() -> Double {
        unitFloat(nextUInt64())
    }
}

/// Non-repeatable randomness from the system generator, for live play where no
/// seed is wanted — the counterpart of the web's `Math.random` default. Uses
/// the same 53-bit construction as `SeededRandom.next()`.
public struct SystemRandomSource: RandomSource {
    private var generator = SystemRandomNumberGenerator()

    public init() {}

    public mutating func next() -> Double {
        unitFloat(generator.next())
    }
}
